File: tools/olchiki_bridge.py
# -*- coding: utf-8 -*-
"""
Ol Chiki to Devanagari pronunciation bridge.

## What this is, and what it is not

This is a **transliteration**, not a translation. It rewrites Santali written in
Ol Chiki into Devanagari letters so a teacher who reads Hindi but not Ol Chiki
can say the line out loud. The meaning is not touched and nothing is
interpreted. The app already has a provenance state for exactly this,
`Provenance.TRANSLITERATED`, labelled "Transliterated, not a translation".

## Why the mapping is derived rather than invented

Ol Chiki letter names in Unicode encode their own phonetic values, in a regular
pattern. The six letters whose names begin with L are the vowels; every other
letter's name is its vowel row followed by its consonant:

    LA  LAA  LI  LU  LE  LO          the six vowels
    AT AG ANG AL                     t  g  ng l
    AAK AAJ AAM AAW                  k  j  m  w
    IS IH INY IR                     s  h  ny r
    UC UD UNN UY                     c  d  nn y
    EP EDD EN ERR                    p  dd n  rr
    OTT OB OV OH                     tt b  v  h

So the table below is read off the Unicode names, not guessed. Verify with
`python tools/olchiki_bridge.py --table`.

## The part that needs a speaker

Ol Chiki is alphabetic: every vowel is a full letter. Devanagari is an abugida:
a consonant carries an inherent 'a' and other vowels attach as matras. Going
from one to the other means deciding when a written Ol Chiki 'a' becomes an
explicit matra and when it is left as the inherent vowel.

This script always writes it explicitly, which is systematic and predictable.
A Santali speaker may prefer the inherent form in some words: the design
mockups render ᱟᱯᱟᱱ as आपन where this produces आपान. Both are readable; only a
speaker can say which a Jharkhand teacher would rather see.

**Treat the output as a draft for review, not as finished content.** It is
labelled TRANSLITERATED on screen for that reason.
"""
import argparse
import io
import json
import sys
import unicodedata

# Independent vowel, and the matra that attaches to a preceding consonant.
VOWELS = {
    'ᱚ': ('ओ', 'ो'),   # LA   o
    'ᱟ': ('आ', 'ा'),   # LAA  aa
    'ᱤ': ('इ', 'ि'),   # LI   i
    'ᱩ': ('उ', 'ु'),   # LU   u
    'ᱮ': ('ए', 'े'),   # LE   e
    'ᱳ': ('ओ', 'ो'),   # LO   o
}

CONSONANTS = {
    'ᱛ': 'त',   # AT   t
    'ᱜ': 'ग',   # AG   g
    'ᱝ': 'ङ',   # ANG  ng
    'ᱞ': 'ल',   # AL   l
    'ᱠ': 'क',   # AAK  k
    'ᱡ': 'ज',   # AAJ  j
    'ᱢ': 'म',   # AAM  m
    'ᱣ': 'व',   # AAW  w
    'ᱥ': 'स',   # IS   s
    'ᱦ': 'ह',   # IH   h
    'ᱧ': 'ञ',   # INY  ny
    'ᱨ': 'र',   # IR   r
    'ᱪ': 'च',   # UC   c
    'ᱫ': 'द',   # UD   d
    'ᱬ': 'ण',   # UNN  nn
    'ᱭ': 'य',   # UY   y
    'ᱯ': 'प',   # EP   p
    'ᱰ': 'ड',   # EDD  dd
    'ᱱ': 'न',   # EN   n
    'ᱲ': 'ड़',  # ERR  rr, Devanagari DDA + nukta
    'ᱴ': 'ट',   # OTT  tt
    'ᱵ': 'ब',   # OB   b
    'ᱶ': 'व',   # OV   v
    'ᱷ': 'ह',   # OH   h
}

# Modifiers and punctuation.
VIRAMA = '्'
MARKS = {
    'ᱸ': 'ं',   # MU TTUDDAG, nasalisation
    'ᱹ': '',         # GAAHLAA TTUDDAAG, glottal. No Devanagari equivalent.
    'ᱺ': 'ं',   # MU-GAAHLAA TTUDDAAG
    'ᱻ': '',         # RELAA
    'ᱼ': '',         # PHAARKAA
    'ᱽ': '',         # AHAD
    '᱾': '।',   # MUCAAD, danda
    '᱿': '॥',   # DOUBLE MUCAAD
}
DIGITS = {chr(0x1C50 + i): chr(0x0966 + i) for i in range(10)}


def transliterate(text):
    """Ol Chiki to Devanagari. Characters outside the block pass through."""
    out = []
    i = 0
    n = len(text)
    while i < n:
        ch = text[i]
        if ch in CONSONANTS:
            out.append(CONSONANTS[ch])
            nxt = text[i + 1] if i + 1 < n else ''
            if nxt in VOWELS:
                out.append(VOWELS[nxt][1])   # matra
                i += 2
                continue
            # A consonant with no vowel after it takes a virama only inside a
            # word. At the end of a word Devanagari convention leaves the final
            # consonant bare, and a trailing virama reads as an error to anyone
            # who knows the script: जोहार्, not जोहार.
            if nxt in CONSONANTS:
                out.append(VIRAMA)
            i += 1
            continue
        if ch in VOWELS:
            out.append(VOWELS[ch][0])        # independent form
            i += 1
            continue
        if ch in MARKS:
            out.append(MARKS[ch])
            i += 1
            continue
        if ch in DIGITS:
            out.append(DIGITS[ch])
            i += 1
            continue
        out.append(ch)
        i += 1
    return ''.join(out)


def print_table():
    print('Ol Chiki letters, with the Unicode name the value is read from:\n')
    for cp in range(0x1C5A, 0x1C78):
        ch = chr(cp)
        name = unicodedata.name(ch).replace('OL CHIKI LETTER ', '')
        if ch in VOWELS:
            print('  %s  %-5s vowel       %s / %s' % (ch, name, VOWELS[ch][0], VOWELS[ch][1]))
        else:
            print('  %s  %-5s consonant   %s' % (ch, name, CONSONANTS.get(ch, '?')))


def main():
    ap = argparse.ArgumentParser()
    ap.add_argument('--table', action='store_true', help='print the derived mapping')
    ap.add_argument('--pack', help='pack json to add a devanagari bridge to')
    ap.add_argument('--write', action='store_true', help='actually write the pack')
    ap.add_argument('--text', help='transliterate one string and exit')
    a = ap.parse_args()

    if a.table:
        print_table()
        return
    if a.text:
        print(transliterate(a.text))
        return
    if not a.pack:
        ap.error('give --table, --text or --pack')

    d = json.load(io.open(a.pack, encoding='utf-8'))
    n = 0

    def walk(node):
        nonlocal n
        if isinstance(node, dict):
            if 'target' in node and isinstance(node['target'], str):
                if any(0x1C50 <= ord(c) <= 0x1C7F for c in node['target']):
                    node['bridge'] = transliterate(node['target'])
                    node['bridgeProvenance'] = 'transliterated'
                    n += 1
            for v in node.values():
                walk(v)
        elif isinstance(node, list):
            for v in node:
                walk(v)

    walk(d.get('entries', {}))
    if a.write:
        io.open(a.pack, 'w', encoding='utf-8', newline='\n').write(
            json.dumps(d, indent=2, ensure_ascii=False, sort_keys=True) + '\n')
        print('wrote %d bridges into %s' % (n, a.pack))
    else:
        print('%d entries would gain a bridge. Re-run with --write.' % n)


if __name__ == '__main__':
    sys.stdout.reconfigure(encoding='utf-8')
    main()

File: app/src/main/java/app/olsaathi/text/OlChikiBridge.kt
package app.olsaathi.text

/**
 * Ol Chiki to Devanagari pronunciation bridge. A port of
 * `tools/olchiki_bridge.py`, and it must stay character for character equal to
 * it: `OlChikiBridgeTest` checks every Santali line in the shipped pack against
 * the `bridge` field that script wrote.
 *
 * This is a transliteration, not a translation. It exists at runtime for one
 * reason: Bhashini's only Santali voice (IIT Madras, `Bhashini/IITM/TTS`) says
 * nothing for Ol Chiki input, 0.02 s of silence with a success response, and
 * speaks the same words written in Devanagari. A live translation arrives in
 * Ol Chiki, so it has to cross this bridge before it can be spoken.
 *
 * Ol Chiki letter names encode their own phonetic values (see the Python file
 * for the derivation). Ol Chiki is alphabetic and Devanagari is an abugida, so
 * a written vowel after a consonant always becomes an explicit matra here.
 */
object OlChikiBridge {

    /** Independent vowel, and the matra that attaches to a preceding consonant. */
    private val VOWELS = mapOf(
        'ᱚ' to ("ओ" to "ो"),   // LA   o
        'ᱟ' to ("आ" to "ा"),   // LAA  aa
        'ᱤ' to ("इ" to "ि"),   // LI   i
        'ᱩ' to ("उ" to "ु"),   // LU   u
        'ᱮ' to ("ए" to "े"),   // LE   e
        'ᱳ' to ("ओ" to "ो"),   // LO   o
    )

    private val CONSONANTS = mapOf(
        'ᱛ' to "त", 'ᱜ' to "ग", 'ᱝ' to "ङ", 'ᱞ' to "ल",
        'ᱠ' to "क", 'ᱡ' to "ज", 'ᱢ' to "म", 'ᱣ' to "व",
        'ᱥ' to "स", 'ᱦ' to "ह", 'ᱧ' to "ञ", 'ᱨ' to "र",
        'ᱪ' to "च", 'ᱫ' to "द", 'ᱬ' to "ण", 'ᱭ' to "य",
        'ᱯ' to "प", 'ᱰ' to "ड", 'ᱱ' to "न", 'ᱲ' to "ड़",
        'ᱴ' to "ट", 'ᱵ' to "ब", 'ᱶ' to "व", 'ᱷ' to "ह",
    )

    private const val VIRAMA = "्"

    private val MARKS = mapOf(
        'ᱸ' to "ं",   // MU TTUDDAG, nasalisation
        'ᱹ' to "",    // GAAHLAA TTUDDAAG, glottal; no Devanagari equivalent
        'ᱺ' to "ं",   // MU-GAAHLAA TTUDDAAG
        'ᱻ' to "",    // RELAA
        'ᱼ' to "",    // PHAARKAA
        'ᱽ' to "",    // AHAD
        '᱾' to "।",   // MUCAAD, danda
        '᱿' to "॥",   // DOUBLE MUCAAD
    )

    /** True when [text] contains any Ol Chiki character at all. */
    fun containsOlChiki(text: String): Boolean = text.any { it.code in 0x1C50..0x1C7F }

    /** Ol Chiki to Devanagari. Characters outside the block pass through. */
    fun toDevanagari(text: String): String {
        val out = StringBuilder(text.length + 8)
        var i = 0
        while (i < text.length) {
            val ch = text[i]
            val consonant = CONSONANTS[ch]
            if (consonant != null) {
                out.append(consonant)
                val next = text.getOrNull(i + 1)
                val matra = next?.let { VOWELS[it] }?.second
                if (matra != null) {
                    out.append(matra)
                    i += 2
                    continue
                }
                // Virama only inside a word: a trailing one reads as an error,
                // जोहार् rather than जोहार.
                if (next != null && next in CONSONANTS) out.append(VIRAMA)
                i += 1
                continue
            }
            val vowel = VOWELS[ch]
            when {
                vowel != null -> out.append(vowel.first)
                ch in MARKS -> out.append(MARKS.getValue(ch))
                ch.code in 0x1C50..0x1C59 -> out.append((0x0966 + (ch.code - 0x1C50)).toChar())
                else -> out.append(ch)
            }
            i += 1
        }
        return out.toString()
    }
}
